### PR TITLE
Exit and liquidation feature

### PR DESCRIPTION
**Pull request by Aerilym**

### CODE REVIEW

File: apps/staking/components/StakedNode/NodeRequestExitButton.tsx [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+    estimateContractWriteFee,
+    simulateStatus,
+    writeStatus,
+    transactionStatus,
+    estimateFeeError,
+    simulateError,
+    writeError,
+    transactionError,
+    simulateEnabled,
+  } = useInitiateRemoveBLSPublicKey({
```
Comment (yougotwill) [inline apps/staking/components/StakedNode/NodeRequestExitButton.tsx:197]: This hook has gotten quite big damn.

File: apps/staking/components/StakedNode/NodeRequestExitButton.tsx [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+};
+
+enum QUERY_STAGE {
+  IDLE,
+  PENDING,
+  SUCCESS,
+  ERROR,
+}
+
+function QueryStatusInformation({
```
Comment (yougotwill) [inline apps/staking/components/StakedNode/NodeRequestExitButton.tsx:342]: Yeah definitely should split these components up a bit

**Review by yougotwill — CHANGES_REQUESTED** [review]
- I can't build the staking app because it's complaining about the `@ts-expect-error` in `tooltip.tsx`. I replaced it with a `@ts-ignore` and it builds
- sent-staking-js/test.ts is missing `contract_id`
- When I went to `/stake` on localhost:3000 I get the following error

```tsx
RangeError: 730.511935977951 can't be converted to BigInt because it isn't an integer
Source

../packages/sent-staking-js/test.ts (37:13) @ BigInt

  35 |   return {
  36 |     address: generateWalletAddress(),
> 37 |     amount: BigInt(Math.random() * 1000),
     |             ^
  38 |     reserved: BigInt(Math.random() * 1000),
  39 |     locked_contributions: [],
  40 |   };
```

File: apps/staking/app/mystakes/modules/ClaimTokensModule.tsx [hunk 3]
```diff
@@ -260,6 +271,16 @@ function ClaimTokensDialog({
 }) {
   const dictionary = useTranslations('modules.claim.dialog');
 
+  const claimRewardsArgs = useMemo(
```
[the remark this answers is not in the document]
Comment (Aerilym) [reply apps/staking/app/mystakes/modules/ClaimTokensModule.tsx:274]: I'm not entirely convinced its needed, but I had to make this change for the new contract call as part of exiting, it probably does nothing in this case but best case scenario its memoised and not triggering extra re-renders, worst case, it does nothing.

File: apps/staking/locales/en.json [hunk 4]
```diff
@@ -413,12 +415,55 @@
       "lastUptimeDescription": "The last time the node provided a proof of availability.",
       "lastRewardHeight": "Last Reward Height: {height}",
       "lastRewardHeightDescription": "Last block height the node received rewards.",
-      "unlockingTimerNotification": "Unlocking {time}",
-      "unlockingTimerDescription": "Time left until the node is unlocked. After unlocking, you can withdraw your stake.",
+      "requestingExitTimerNotification": "Exit available {time}",
+      "requestingExitDescription": "Time left until the node is available for exit. Return {relative_time} ({date}) to exit the node.",
```
[the remark this answers is not in the document]
Comment (Aerilym) [reply apps/staking/locales/en.json:419]: yeah `{date}` is, the dynamic variable `{date}` is just being wrapped in parenthesis to make the string look like:
`Time left until the node is available for exit. Return in 14 days (Wednesday, 11 September 2024 at 11:23:52pm GMT+10) to exit the node.`

File: apps/staking/components/StakedNode/NodeRequestExitButton.tsx [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+          >
+            {dictionary('buttons.cancel')}
+          </Button>
+        </AlertDialogCancel>
+      </AlertDialogFooter>
+    </>
+  );
+}
+
+function RequestNodeExitContractWriteDialog({ node }: { node: StakedNode }) {
```
[the remark this answers is not in the document]
Comment (Aerilym) [reply apps/staking/components/StakedNode/NodeRequestExitButton.tsx:178]: Yeah im looking to make a contract write dialog generic as this has quite similar code to 2 other places already

File: apps/staking/components/StakedNode/NodeRequestExitButton.tsx (hunk 2 above)
Comment (Aerilym) [reply apps/staking/components/StakedNode/NodeRequestExitButton.tsx:342]: Yeah for sure, see above

File: packages/contracts/hooks/ServiceNodeRewards.tsx [hunk 6]
```diff
@@ -210,3 +210,29 @@ export function useAddBLSPubKey({
...
+};
+
+export function useInitiateRemoveBLSPublicKey({
+  contractId,
+}: {
+  contractId: number;
+}): UseInitiateRemoveBLSPublicKeyReturn {
+  const chain = useChain();
+
+  const defaultArgs = useMemo(() => [BigInt(contractId ?? 0)] as [bigint], [contractId]);
```
[the remark this answers is not in the document]
Comment (Aerilym) [reply packages/contracts/hooks/ServiceNodeRewards.tsx:225]: It has to be `[bigint]` as that is the smart contract arg, as in, it is a literal single item array where that one item is a `bigint`. This isn't be the same as `Array<bigint>` which is a 0-∞ size array where each item is a `bigint`.

File: apps/staking/components/StakedNodeCard.tsx [hunk 7]
```diff
@@ -103,12 +119,16 @@ const isUnlocked = (node: StakedNode): node is UnlockedStakedNode =>
  * @returns `true` if the node is being deregistered, `false` otherwise.
  */
 const isBeingDeregistered = (node: StakedNode): node is DecommissionedStakedNode =>
-  node.state === NODE_STATE.DECOMMISSIONED;
+  !!(
+    node.state === NODE_STATE.DECOMMISSIONED &&
+    'deregistrationDate' in node &&
+    node.deregistrationDate
```
[the remark this answers is not in the document]
Comment (Aerilym) [reply apps/staking/components/StakedNodeCard.tsx:125]: `'deregistrationDate' in node` checks that `deregistrationDate` is a property on the `node` object, as in it checks if it is declared/exists. This then allows us to do `node.deregistrationDate` as the previous check was true so we know `deregistrationDate` is declared and we can now check for its thuthyness

File: packages/ui/components/ui/tooltip.tsx [hunk 8]
```diff
@@ -9,6 +9,7 @@ import {
   type ReactNode,
   useState,
 } from 'react';
+/** @ts-expect-error TS doesnt know what its talking about */
 import { useDebounce } from '@uidotdev/usehooks';
```
[the remark this answers is not in the document]
Comment (Aerilym) [reply packages/ui/components/ui/tooltip.tsx:13]: damn, ok fixed

**Comment by Aerilym** [comment]
> * I can't build the staking app because it's complaining about the `@ts-expect-error` in `tooltip.tsx`. I replaced it with a `@ts-ignore` and it builds
> 
>     * sent-staking-js/test.ts is missing `contract_id`
> 
>     * When I went to `/stake` on localhost:3000 I get the following error
> 
> 
> ```tsx
> RangeError: 730.511935977951 can't be converted to BigInt because it isn't an integer
> Source
> 
> ../packages/sent-staking-js/test.ts (37:13) @ BigInt
> 
>   35 |   return {
>   36 |     address: generateWalletAddress(),
> > 37 |     amount: BigInt(Math.random() * 1000),
>      |             ^
>   38 |     reserved: BigInt(Math.random() * 1000),
>   39 |     locked_contributions: [],
>   40 |   };
> ```

Changes made. Ive also fixed those broken tests for you :)

File: packages/contracts/hooks/ServiceNodeRewards.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (yougotwill) [reply packages/contracts/hooks/ServiceNodeRewards.tsx:225]: I see. My mistake sorry. All good 👍

File: apps/staking/components/StakedNodeCard.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (yougotwill) [reply apps/staking/components/StakedNodeCard.tsx:125]: Cool

**Review by yougotwill — APPROVED** [review]
LGTM 🚢

**Pull request merged**